CRIU supports Java debugger via the restore option file

JasonFengJ9 opened the pull request:
CRIU supports Java debugger via the restore option file

Debugger support related code are guarded with `isDebugOnRestoreEnabled()`;
For `-Xint` mode, do not disable `JVMTI` capabilities required for `JDWP` debugger before checkpoint, add a capability `can_access_local_variables`;
For `JIT` mode, add `JVMTI` capabilities required for `JDWP` debugger before checkpoint;
Hooked `J9HOOK_VM_PREPARING_FOR_RESTORE` event to determine if a `JDWP` agent is specified in the restore option file, also hooked `J9HOOK_VM_CRIU_RESTORE` to load the agent libraries;
Refactored the agent library creation from `J9VMInitArgs`;
Added some trace points.

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

Closes https://github.com/eclipse-openj9/openj9/issues/16959

JasonFengJ9 commented:
FYI @tajila

JasonFengJ9 commented:
The CRIU JDWP debugger via the restore option file requires both extension repo PR https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/591 and its backports (in progress) and this PR.
However, this PR alone won't break any existing builds such as Java 8 that won't be backported.

JasonFengJ9 commented:
FYI @dsouzai 
This PR has the following messages post-restore when no `JDWP` agent is specified in the restore option file.
```
JVMJITM044W Some or all compiled code in the code cache invalidated post restore.
JVMJITM043W AOT load and compilation disabled post restore.
```
which is expected to be resolved via a JIT change like the patch you provided.

Note: this `JDWP` debugger support via the option file is only available with `-XX:+DebugOnRestore`.

dsouzai commented:
I was reading over https://github.com/eclipse-openj9/openj9/issues/18866, specifically the VM Coordination section, and wanted to check if Points 1 and 2 are addressed, namely:

1. Decompile immediately on restore if a user enables debug post-restore
2. If debug is not enabled post-restore, but normal HCR is possible, then the VM needs to ensure that stacks with FSD bodies get decompiled.

Regarding 1, although currently (i.e. with this PR and https://github.com/eclipse-openj9/openj9/pull/20047) we will invalidate all compiled code if debug is enabled post-restore, it is still possible for JIT'd code to execute, namely:
* JNI Thunks
* Failed proactive compilation

For a failed proactive compilation, we will fall back to using the FSD body which means that a thread will continue to execute that body. Therefore, the VM needs to decompile immediately on restore if debug is enabled. Now, once I add the ability to reuse the FSD code we generate pre-checkpoint this won't be necessary, but for now I believe we need to close any gaps.

In the case of JNI Thunks, I'm working on a PR for that.

I think once we address all of this, debug on restore should be completely functional (if not optimal). FYI @TobiAjila @vijaysun-omr

dsouzai commented:
>For a failed proactive compilation, we will fall back to using the FSD body which means that a thread will continue to execute that body. Therefore, the VM needs to decompile immediately on restore if debug is enabled. Once we have the ability to reuse the FSD code we generate pre-checkpoint this won't be necessary, but for now I believe we need this to make debug fully functional.

Hm actually... I guess it's not actually a problem if it's running FSD code because that is already set up to deal with debug events. So I guess I just need to address the JNI Thunk issue; the rest is covered by the method body invalidation that already occurs.

I'll think about this some more, but now I don't believe we need to decompile immediately anymore heh.

dsouzai commented:
> In the case of JNI Thunks, I'm working on a PR for that.

Opened https://github.com/eclipse-openj9/openj9/pull/20108.

JasonFengJ9 commented on runtime/jvmti/jvmtiCapability.c:
```diff
@@ -540,55 +546,74 @@ mapCapabilitiesToEvents(J9JVMTIEnv * j9env, jvmtiCapabilities * capabilities, J9
-
-	if (capabilities->can_generate_frame_pop_events) {
-		rc |= eventHookFunction(j9env, JVMTI_EVENT_FRAME_POP);
+	/* can't skip for criu */
```
The `eventHookFunction()` is `reserveEvent` which invokes `J9HookReserve()` later https://github.com/eclipse-openj9/openj9/blob/c24c9542ffd415e5431ae2a2109129f6c55e4268/runtime/jvmti/jvmtiCapability.c#L414


Those capabilities have to be reserved if requested via `jvmtiAddCapabilities()` otherwise they will be disabled by JIT, for example, `can_generate_method_exit_events/JVMTI_EVENT_METHOD_EXIT` related hooks are `J9HOOK_VM_METHOD_RETURN/J9HOOK_VM_NATIVE_METHOD_RETURN` https://github.com/eclipse-openj9/openj9/blob/c24c9542ffd415e5431ae2a2109129f6c55e4268/runtime/jvmti/jvmtiHook.c#L692-L694
The hook `J9HOOK_VM_METHOD_RETURN` is to be disabled if not reserved in advance
https://github.com/eclipse-openj9/openj9/blob/c24c9542ffd415e5431ae2a2109129f6c55e4268/runtime/compiler/control/J9Options.cpp#L3169


Similarly for the other following capabilities, if they are not reserved and a `jdwp` agent is specified at the restore option file, the capability in question can't be added and the debugger won't be able to run.

tajila replied:
okay thanks

Please add some of that detail in the comment

JasonFengJ9 replied:
Yup, added a comment.

JasonFengJ9 commented:
Rebased to resovle the merge conflict.

tajila commented on runtime/jvmti/jvmtiHook.c:
```diff
@@ -539,6 +556,21 @@ jvmtiHookVMCheckpoint(J9HookInterface **hook, UDATA eventNum, void *eventData, v
+
+	Trc_JVMTI_jvmtiHookVMRestore_Entry();
+
+	data->currentThread->javaVM->internalVMFunctions->internalExitVMToJNI(data->currentThread);
```
This will be called while we are holding on to exclusive or safe point access. Im not sure this operation is safe.

D we need to release VM access

JasonFengJ9 replied:
`criuRestoreInitializeLib()` eventually releases VM access when it loads the agent library.
```
		vm->internalVMFunctions->internalEnterVMFromJNI(currentThread);
		vm->internalVMFunctions->acquireExclusiveVMAccess(currentThread);
...
		vm->internalVMFunctions->releaseExclusiveVMAccess(currentThread);
		vm->internalVMFunctions->internalExitVMToJNI(currentThread);
#0  allocateEnvironment (invocationJavaVM=0x7ffff0283af0, version=806682624, penv=0x7fffe0830aa0 <s>) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiHelpers.cpp:382
#1  0x00007ffff56e4e70 in jvmtiHookGetEnv (hook=0x7ffff0012120, eventNum=18, eventData=0x7ffff7d35ac0, userData=0x7ffff00e9960) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiHook.c:1969
#2  0x00007ffff7f9c6ce in J9HookDispatch (hookInterface=0x7ffff0012120, taggedEventNum=18, eventData=0x7ffff7d35ac0) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/omr/util/hookable/hookable.cpp:237
#3  0x00007ffff79d5eca in GetEnv (jvm=0x7ffff0283af0, penv=0x7fffe0830aa0 <s>, version=806682624) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/vm/jniinv.c:898
#4  0x00007fffe07ec1f5 in Agent_OnLoad (vm=0x7ffff0283af0, options=0x7ffff0296b75 "transport=dt_socket,server=y,suspend=y", reserved=0x0) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:208
#5  0x00007ffff56f8565 in issueAgentOnLoadAttach (vm=0x7ffff000fdc0, agentLibrary=0x7ffff00e9b98, options=0x7ffff0296b75 "transport=dt_socket,server=y,suspend=y", loadFunctionName=0x7ffff573255e "Agent_OnLoad", foundLoadFn=0x7ffff7d35c88)
    at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiStartup.c:519
#6  0x00007ffff56f8cac in loadAgentLibraryGeneric (vm=0x7ffff000fdc0, agentLibrary=0x7ffff00e9b98, loadFunctionName=0x7ffff573255e "Agent_OnLoad", loadStatically=0, found=0x7ffff7d35c88, errorMessage=0x7ffff7d35c90)
    at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiStartup.c:632
#7  0x00007ffff56f9ae1 in loadAgentLibrary (vm=0x7ffff000fdc0, agentLibrary=0x7ffff00e9b98) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiStartup.c:862
#8  0x00007ffff56f79e9 in processAgentLibraryFromArgsList (vm=0x7ffff000fdc0, argsList=0x7ffff027d9f0, agentColon=0x7ffff573227b "-agentlib:", loadLibrary=1) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiStartup.c:190
#9  0x00007ffff56f7d2e in criuRestoreInitializeLib (vm=0x7ffff000fdc0, j9env=0x7ffff00e9960) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiStartup.c:278
#10 0x00007ffff56dff07 in jvmtiHookVMRestoreCRIUInit (hook=0x7ffff0012120, eventNum=103, eventData=0x7ffff7d35fe0, userData=0x7ffff00e9960) at /root/builds/jdk21-jdwpoptionfile-0904/openj9-openjdk-jdk21/openj9/runtime/jvmti/jvmtiHook.c:568
```

`criuRestoreInitializeLib()` is registered with `J9HOOK_VM_PREPARING_FOR_RESTORE` to ensure it is invoked before JIT hooks, and determine if a `JDWP` agent is specified in the restore option file.
`TRIGGER_J9HOOK_VM_PREPARING_FOR_RESTORE()` is fired before https://github.com/eclipse-openj9/openj9/blob/bfa3cfd037f9afb183486dd048bd9a9bf016d1e2/runtime/vm/CRIUHelpers.cpp#L1943

I am going to keep the `J9HOOK_VM_PREPARING_FOR_RESTORE` hook to process the restore option arguments and enable `J9VM_CRIU_IS_JDWP_ENABLED` accordingly, and move the actual agent loading after `releaseSafeOrExcusiveVMAccess()`.

dsouzai replied:
> to ensure it is invoked before JIT hooks

The JIT also registers against the `J9HOOK_VM_PREPARING_FOR_RESTORE` hook; is it a guarantee that the jit hook will be called after `criuRestoreInitializeLib`?

JasonFengJ9 replied:
Yes, with `J9HOOK_AGENTID_FIRST` at `J9HookRegisterWithCallSite(vmHook, J9HOOK_TAG_AGENT_ID | J9HOOK_VM_PREPARING_FOR_RESTORE, jvmtiHookVMRestoreCRIUInit, OMR_GET_CALLSITE(), jvmtiData, J9HOOK_AGENTID_FIRST)`.

JasonFengJ9 replied:
Hooked `J9HOOK_VM_PREPARING_FOR_RESTORE` event to determine if a `JDWP` agent is specified in the restore option file, also hooked `J9HOOK_VM_CRIU_RESTORE` to load the agent libraries which is triggered at https://github.com/eclipse-openj9/openj9/blob/4fbb9648b26ac6a2d5df22973d9b8feeeab580f3/runtime/vm/CRIUHelpers.cpp#L1943-L1944

@tajila this is ready for another look.

tajila commented:
same issue
```
JVMJITM043W AOT load and compilation disabled post restore.
post restore hook
JVMJ9VM007E Command-line option unrecognised: -agentlib:jdwp=transport=dt_socket,server=y,suspend=y
Exception in thread "main" org.eclipse.openj9.criu.JVMRestoreException: The JVM could not enable all the restore options specified
	at openj9.criu/org.eclipse.openj9.criu.CRIUSupport.checkpointJVM(CRIUSupport.java:534)
	at Debug.checkPointJVM(Debug.java:32)
	at Debug.main(Debug.java:16)
Caused by: openj9.internal.criu.JVMRestoreException: The JVM could not enable all the restore options specified
	at java.base/openj9.internal.criu.InternalCRIUSupport.checkpointJVMImpl(Native Method)
	at java.base/openj9.internal.criu.InternalCRIUSupport.checkpointJVM(InternalCRIUSupport.java:997)
	at openj9.criu/org.eclipse.openj9.criu.CRIUSupport.checkpointJVM(CRIUSupport.java:530)
	... 2 more
```

tajila commented:
@JasonFengJ9 We will need to add a docs issue for this as well

JasonFengJ9 commented:
> -Xrunjdwp:transport=dt_socket,address=9999,server=y,suspend=y

Added support of `-Xrunjdwp`.

@tajila this is ready for another look.

tajila commented:
jenkins test sanity xlinux jdk17

tajila commented:
jenkins test sanity alinux64 jdk21

tajila commented:
jenkins test sanity win jdk8

JasonFengJ9 commented:
Fixed the PR test failure due to an earlier `break` for undecorated `JDWP` agent processing while multiple agents were present. 

In addition, disabled following failure conditions within `criu_jitPostRestore.xml`:
```
		<!-- output type="failure" caseSensitive="yes" regex="no">Some or all compiled code in the code cache invalidated post restore.</output>
		<output type="failure" caseSensitive="yes" regex="no">JIT compilation disabled post restore.</output -->
```
Verified manually that these failure conditions can re-enabled with 
* https://github.com/eclipse-openj9/openj9/pull/20047 

Also moved `DisposeEnvironment(jvmti_env)` into `jvmtiHookVMRestoreStartAgent()` which is the last part of the cleanup if there was no JDWP agent specified. It releases VM access and hence can't be invoked within criuDisableHooks() from `J9HOOK_VM_PREPARING_FOR_RESTORE`.

@tajila this is ready for another look.

tajila commented:
jenkins test sanity win jdk8

tajila commented:
jenkins test sanity alinux64 jdk21

tajila commented:
jenkins test sanity xlinux64 jdk17

tajila commented:
jenkins test sanity xlinux jdk17
